connection/TypeDBTransactionImpl.java

```diff
@@ -109,17 +110,23 @@ private Res execute(Req.Builder request, boolean batch) {
     }
 
     private QueryFuture<Res> query(Req.Builder request, boolean batch) {
-        if (!isOpen()) throw new TypeDBClientException(TRANSACTION_CLOSED);
+        if (!isOpen()) drainAndThrow();
         BidirectionalStream.Single<Res> single = bidirectionalStream.single(request, batch);
         return single::get;
     }
 
     @Override
     public Stream<ResPart> stream(Req.Builder request) {
-        if (!isOpen()) throw new TypeDBClientException(TRANSACTION_CLOSED);
+        if (!isOpen()) drainAndThrow();
         return bidirectionalStream.stream(request);
     }
 
+    private void drainAndThrow() {
+        if (bidirectionalStream.hasErrors()) {
+            throw new TypeDBClientException(TRANSACTION_CLOSED_WITH_ERRORS, bidirectionalStream.drainErrors());
+        } else throw new TypeDBClientException(TRANSACTION_CLOSED);
+    }
```

Comment:
drain the errors from the queue and throw them all

Reply:
I don't think this method name is intuitive - how about `throwTransactionClosedError()`? It's scoped to the class, and when looking at the class as a whole, that name makes the purpose clearer.

Also, `hasErrors` feels unnecessary here (looking at its implementation, too, it seems largely copied from/to `drainErrors`); I think we can delete `hasErrors` and write
```java
List<StatusRuntimeException> errors = bidirectionalStream.drainErrors();
if (errors.isEmpty()) throw new TypeDBClientException(TRANSACTION_CLOSED);
else throw new TypeDBClientException(TRANSACTION_CLOSED_WITH_ERRORS, errors);
```

Reply:
good ideas